client/core/trade.go

```diff
@@ -1224,6 +1257,83 @@ func (t *trackedTrade) finalizeRedeemAction(match *matchTracker, coinID []byte)
 	return errs.ifAny()
 }
 
+func (t *trackedTrade) beginFindRedemption(match *matchTracker) {
+	swapCoinID := dex.Bytes(match.MetaData.Proof.TakerSwap)
+	fromAsset := t.wallets.fromAsset.ID
+	resultChan, err := t.wallets.fromWallet.FindRedemption(swapCoinID)
+	if err != nil {
+		log.Errorf("failed to start redemption search for match %v, %s contract %s: %v",
+			match.id, unbip(fromAsset), coinIDString(fromAsset, swapCoinID), err)
+		return
+	}
+	// Redemption search started, start goroutine to monitor result channel.
+	t.awaitingRedemption[swapCoinID.String()] = match.id
+	go t.waitForRedemption(resultChan)
+}
+
+// waitForRedemption waits on the passed `redemptionChan` for FindRedemption
+// results. Should be run as a goroutine.
+func (t *trackedTrade) waitForRedemption(redemptionChan chan *asset.FindRedemptionResult) {
+	frr := <-redemptionChan
+	if frr == nil {
+		return // channel closed without data
+	}
+
+	fromAsset := t.wallets.fromAsset
+	var match *matchTracker
+
+	t.mtx.Lock()
+	orderID, found := t.awaitingRedemption[frr.ContractCoinID.String()]
+	if found {
+		delete(t.awaitingRedemption, frr.ContractCoinID.String())
+		match, found = t.matches[orderID]
+	}
+	if !found {
+		t.mtx.Unlock()
+		log.Warnf("received find redemption result for unknown contract coin (%s: %s) on order %s",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID())
+		return
+	}
+
+	if frr.Err != nil {
+		t.mtx.Unlock()
+		log.Errorf("FindRedemption error: coin (%s: %s), order %s, match %s",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID(), match.id)
+		return
+	}
+
+	if match.Match.Status != order.TakerSwapCast {
+		t.mtx.Unlock()
+		log.Errorf("received find redemption result at wrong step, order %s, match %s, status %s",
+			t.ID(), match.id, match.Match.Status)
+		return
+	}
+
+	_, _, proof, _ := match.parts()
+	// TODO: Is this validation necessary?
+	if !t.wallets.toWallet.ValidateSecret(frr.Secret, proof.SecretHash) {
+		t.mtx.Unlock()
+		log.Errorf("FindRedemption error: coin (%s: %s), order %s, match %s, invalid secret %s, hash %s",
+			fromAsset.Symbol, coinIDString(fromAsset.ID, frr.ContractCoinID), t.ID(), match.id,
+			frr.Secret, proof.SecretHash)
+		return
+	}
```

Comment:
> Is this validation necessary?

In theory, no. But let's not trust the `Wallet` to do it right in `FindRedemption`.